### ISSUE TITLE
fp.spacer.gpdr fp.spacer.mbqi HORN Assertion violation at ../src/muz/spacer/spacer_context.cpp Line: 3502

### ISSUE THREAD

**Issue by muchang** [Z3Prover/z3#3892]
Hi,
For this formula:
```
(set-logic HORN)
(set-option :fp.spacer.gpdr true)
(set-option :fp.spacer.mbqi false)
(assert (exists ((a Int)) (= a 0)))
(check-sat)
```
Z3  throws out an assertion violation:
```
ASSERTION VIOLATION
File: ../src/muz/spacer/spacer_context.cpp
Line: 3502
m_weak_abs
(C)ontinue, (A)bort, (S)top, (T)hrow exception, Invoke (G)DB
```
When replacing  (= a 0) with (= a a), the assertion violation disappears. 

OS: Ubuntu 18.04
Commit: 2673807

**Comment by rainoftime** [comment]
https://github.com/Z3Prover/z3/issues/3780

**Comment by andrewvaughanj** [comment]
@rainoftime your comment seems weird; you referred to @agurfinkel's comment of:

> Non-linear arithmetic is not supported in spacer.

which this instance does not infringe on.

Ignoring the quantifier (which can be written as a `declare-const` rather than the line `exists`), this example is in QF_IDL.

**Comment by agurfinkel** [comment]
@rainoftime the issue you reference explains that this is not a bug and asks you not to report more issues like this one.

Incomplete solver will return unknown, in debug mode, I check for this with an assert. In release mode, spacer propagates this unknown up. 

How is this an issue?

It would help if when reporting assertion failures in debug builds you also explain why you believe this is a violation of the spec.

**Comment by andrewvaughanj** [comment]
@agurfinkel: this issue wasn't raised by @rainoftime but by @muchang -- so I guess @rainoftime was referring @muchang to that other issue to say "this shouldn't be reported" (however, just providing the issue number made that a bit confusing!).

**Comment by zhendongsu** [comment]
@agurfinkel  
Hi Arie: Your explanation above is much clearer than the one given in  https://github.com/Z3Prover/z3/issues/3780, based on which we thought that this issue might have been of interest as the test does not involve non-linear arithmetic.  

In fact, it would be useful (which does not seem too much work either) to provide two pieces of quick information if you could: 
(1) Which options for spacer are meant for internal consumption? 
(2) When raising an assertion like the one above, it might be useful to comment that it is internal and for debugging purposes. 

Thank you.

**Comment by agurfinkel** [comment]
That is easy: all assertions are for internal consumption only to be used for debugging. They are all implicitly commented as such. I never expect users to run the tool in debug mode. I use assertions to debug segmentation faults and other failures to zoom on what implicit assumptions are violated by the input.

Options are harder. All options are for internal consumption. The user is expected to run the tool without options. This , however, does not always work best. Options are added to support specific cases that users have.  Each option is added for a particular use case and supports this use case only. Best way to see what options are used is by looking at public examples. Say this one: https://spacerexamples-ariegurfinkel.notebooks.azure.com/j/notebooks/Dagstuhl2019.ipynb

From which, the only available option is `fp.engine`.

Everything is designed to support rapid prototyping. This means that there are many options, but most of them are not compatible with one another. When the user applies incompatible set of options, the behavior is undefined: assertion, segfault, wrong answer -- these are all expected and specified behaviors in that case. These should not be reported -- they satisfy the spec, they are not issues!

I apologize for being a bit harsh. Unfortunately, I don't have time this week to look at all the spacer related issues. However, all the ones that I have seen so far are noise. The ones that I see @NikolajBjorner fix, are fixed by hiding useful experimental code further in the tool. The end result is that features that are not yet ready for primetime are now harder to work on. I have to rely on my custom fork instead of the master branch to hide my code from this noise. This is quite counter-productive. I would rather spend my time on developing better solving algorithms then think of better ways to make experimental code harder to reach for random testers while still being accessible to users.

**Comment by zhendongsu** [comment]
> However, all the ones that I have seen so far are noise.

Arie, I assume that you meant the assertion violations, not soundness or invalid model bugs like, for example: 
https://github.com/Z3Prover/z3/issues/3890
https://github.com/Z3Prover/z3/issues/3858
https://github.com/Z3Prover/z3/issues/3872

But, even for assertion violations, how about https://github.com/Z3Prover/z3/issues/3849 and https://github.com/Z3Prover/z3/issues/3863 for example?

**Comment by agurfinkel** [comment]
Mostly noise, except for one that I have to look deeper. I suspect the one is noise as well, but that one I have to look at. Here are the explanations:

#3890 -- probably noise. Not in HORN fragment.
#3858 -- maybe a bug in elim_term_ite. This one will be on my todo list
#3872 -- noise. uninterpreted functions are not supported by HORN
#3849 -- noise. was fixed by early termination.
#3863 -- noise. UF in HORN. highly experimental option.

**Comment by agurfinkel** [comment]
It would be greatly appreciated if you can triage the bugs and only report things that are genuine issues, 
or high severity like memory corruption and segmentation faults.

**Comment by zhendongsu** [comment]
> It would be greatly appreciated if you can triage the bugs and only report things that are genuine issues,
> or high severity like memory corruption and segmentation faults.

Arie, we do spend quite some careful effort to report only interesting, genuine issues --- to the best of our knowledge.  I believe what constitutes a (good/interesting) bug is highly subjective, and this becomes even trickier from considering the views of developers and users.  As it is clear, you and Nikolaj may also not always agree on what is an issue/non-issue (in the context of spacer).

**Comment by agurfinkel** [comment]
You are filing issues in experimental code using hidden parameters. It is up to you to justify an issue. One or two such issues for developers to triage is reasonable, but a constant stream is hard to deal with.

I have not discussed this with @NikolajBjorner , he might have a different opinion, but what I see from his commits are just ways to make experimental code harder to access. Granted that this is end of term for me and I have very few cycles to keep track of what is happening. There might be genuinely interesting issues. But as I said, all issues that I have seen so far, and all fixes from @NikolajBjorner in spacer that I kept track of are noise.

**Comment by zhendongsu** [comment]
> You are filing issues in experimental code using hidden parameters. It is up to you to justify an issue. One or two such issues for developers to triage is reasonable, but a constant stream is hard to deal with.

To be clear, Arie, I only refer to ones filed by @muchang and @wintered.  We generally focus on genuine soundness and invalid model bugs, not crashes and such (we find them as by-products). 

We appreciate the great and continued work of folks like Nikolaj and the CVC4 crew (and you) on making modern SMT solvers more powerful and practically usable.  We try to help make them more reliable and robust.

**Comment by zhendongsu** [comment]
Arie, I think one needs to be careful here with the definition of "noise".   These are certainly genuine bugs, although might not be due to spacer.  The following is a clear example: 

> #3872 -- noise. uninterpreted functions are not supported by HORN

**Comment by agurfinkel** [comment]
I apologize for my use of words -- "noise" was a bad choice.

A behavior is a bug if it does not correspond to the specification. The specification for spacer says that in presence of UF and for non-HORN formulas its behavior is undefined. 

As in C standard, undefined behavior in the specification is done to make implementation more flexible and easier to maintain. 

The issue exposes this part of the specification. 

I can change the specification, for example, by requiring to return an error message instead of undefined behavior. However, I find changing specification in this way counter-productive. 

With all of that said, I agree that undefined behavior in #3872 is unfortunate. It is better if spacer returns `unknown` in this case. As I said, I looked only a small fraction of the issues in depth. I will look at them more in coming weeks. Right now, I react only to indicate on features that are not supported. Please do not create more HORN examples that are outside of HORN fragment, until the existing ones are fixed.

**Comment by NikolajBjorner** [comment]
> With all of that said, I agree that undefined behavior in #3872 is unfortunate. 

It returns unknown now. 
There is a "black-list" feature that filters out known incompatibilities as encoded in muz/base/rule_properties. 
It is for sure much more fragile than a stringent white-list, but could give a way to learn.

Some assertions can be reasonably replaced by exceptions.

You can add back assertion in dl_mk_elim_term_ite. I revised it a bit this morning to fix #3858.
The example also provides an example where a top-level ite is not removed.
It doesn't look right to use the elim_term_ite_tactic because it could introduce auxiliary names that are functions.

**Comment by agurfinkel** [comment]
I will start looking at this next week. Assigned the issue to myself to not loose it.

We need a macro to wrap exceptions that are assertions. This way, in Debug mode the macro can expand to exception and in Developer mode it can expand to an assertion.

In general, I would be careful from introducing too many dependencies on exceptions in the code. My understanding that architectures like wasm don't support them. Being able to compile Z3 with assertions disabled (`-fnoexceptions`) might be a useful option.
